### PR TITLE
Fix segfaults in RTCDataChannel and RTCPeerConnection by not freeing objects until destruction

### LINKED ISSUE #236: Datachannel onclose event does not trigger
I do not get the onclose event from datachannels. I've tested this with the latest 4 versions and both on Linux (Ubuntu 14.04) and OS X (Maverick). I've tested with both the example code and custom code but the onclose event is not triggered. Have I missed something here?

### LINKED ISSUE #325: segfault when client disconnects
I detect client disconnects through a websocket and try to stop sending data asap. Either way wrtc should not segfault.

Extract from my own log:

...
sending frame
sending frame
sending frame
sending frame
sending frame
sending frame
client closed
PID 21292 received SIGSEGV for address: 0x0
/home/zubzub/git/greenfield/node_modules/segfault-handler/build/Release/segfault-handler.node(+0x1bb8)[0x7fc436dfbbb8]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x11390)[0x7fc4474da390]
/home/zubzub/git/greenfield/node_modules/wrtc/build/wrtc/v0.0.62/Release/node-v57-linux-x64/wrtc.node(_ZN11node_webrtc11DataChannel13GetReadyStateEN2v85LocalINS1_6StringEEERKN3Nan20PropertyCallbackInfoINS1_5ValueEEE+0x54)[0x7fc4440d7bf4]
/home/zubzub/git/greenfield/node_modules/wrtc/build/wrtc/v0.0.62/Release/node-v57-linux-x64/wrtc.node(+0x7f1c7)[0x7fc4440d71c7]
[0x38ee83d51080]

### PR DESCRIPTION
**Pull request by nazar-pc**
This fixes #236 and fixes #325.

Hopefully, this can be merged and released soon.

P.S. I'm not a C/C++ developer even remotely and might have done something horribly wrong, but it seems to work correctly.

### CODE REVIEW

**Comment by markandrus** [comment]
Thanks, @nazar-pc! Do you know why we were null-ing out the `_jinglePeerConnection` and related objects? I thought there was a performance reason. Not sure, but `git blame` should show. I’m AFK at the moment, otherwise would check.

**Comment by nazar-pc** [comment]
Interesting, after reading #291 I've reworked PR into simpler one.
BTW, tests from #291 are failing as of current `develop`, but they work properly with this PR.

File: src/datachannel.cc [hunk 1]
```diff
@@ -229,37 +229,39 @@ NAN_METHOD(DataChannel::Send) {
 
   DataChannel* self = Nan::ObjectWrap::Unwrap<DataChannel>(info.This());
 
-  if (info[0]->IsString()) {
-    Local<String> str = Local<String>::Cast(info[0]);
-    std::string data = *String::Utf8Value(str);
+  if (self->_jingleDataChannel != nullptr) {
```
Comment (markandrus) [inline src/datachannel.cc:232]: In a future PR, we should actually throw an InvalidStateError. See [here](https://w3c.github.io/webrtc-pc/#dfn-send()).

File: src/datachannel.cc [hunk 2]
```diff
@@ -293,7 +298,13 @@ NAN_GETTER(DataChannel::GetBufferedAmount) {
...
   DataChannel* self = Nan::ObjectWrap::Unwrap<DataChannel>(info.Holder());
 
-  uint64_t buffered_amount = self->_jingleDataChannel->buffered_amount();
+  uint64_t buffered_amount;
+
+  if (self->_jingleDataChannel != nullptr) {
+    buffered_amount = self->_jingleDataChannel->buffered_amount();
+  } else {
+    buffered_amount = 0;
+  }
```
Comment (markandrus) [inline src/datachannel.cc]: WDYT about writing this as

```c++
uint64_t buffered_amount = self->_jingleDataChannel
  ? self->_jingleDataChannel->buffered_amount()
  : 0;
```

Also, in the future, this property should behave differently. See [here](http://w3c.github.io/webrtc-pc/#dom-rtcdatachannel-bufferedamount):

> the attribute does not reset to zero once the channel closes

That being said, the point of this PR is just to fix a segfault, not enforce spec-compliance, so we should punt this to a future PR.

File: src/datachannel.cc [hunk 3]
```diff
@@ -304,7 +315,13 @@ NAN_GETTER(DataChannel::GetLabel) {
...
   DataChannel* self = Nan::ObjectWrap::Unwrap<DataChannel>(info.Holder());
 
-  std::string label = self->_jingleDataChannel->label();
+  std::string label;
+
+  if (self->_jingleDataChannel != nullptr) {
+    label = self->_jingleDataChannel->label();
+  } else {
+    label = "";
+  }
```
Comment (markandrus) [inline src/datachannel.cc]: Ditto re: ternary as above.

File: src/datachannel.cc [hunk 4]
```diff
@@ -315,7 +332,13 @@ NAN_GETTER(DataChannel::GetReadyState) {
...
   DataChannel* self = Nan::ObjectWrap::Unwrap<DataChannel>(info.Holder());
 
-  webrtc::DataChannelInterface::DataState state = self->_jingleDataChannel->state();
+  webrtc::DataChannelInterface::DataState state;
+
+  if (self->_jingleDataChannel != nullptr) {
+    state = self->_jingleDataChannel->state();
+  } else {
+    state = webrtc::DataChannelInterface::kClosed;
+  }
```
Comment (markandrus) [inline src/datachannel.cc]: Ditto re: ternary as above.

File: src/datachannel.cc (hunk 1 above)
Comment (markandrus) [reply src/datachannel.cc:232]: https://github.com/js-platform/node-webrtc/issues/343

**Comment by markandrus** [comment]
I left a few comments. Please let me know what you think; otherwise, the changes look good to me. Please also add yourself to the `AUTHORS` file if you’d like.

**Comment by nazar-pc** [comment]
Refactored as suggested, comments make perfect sense to me.

**Comment by markandrus** [comment]
Thanks! @nazar-pc will try to get a release out in next few days

**Comment by 3mcd** [comment]
Great work!

**Pull request merged**